lib/spack/spack/test_state.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from types import ModuleType
+
+import pickle
+import pydoc
+import io
+import sys
+import multiprocessing
+
+import spack.architecture
+import spack.config
+
+
+patches = None
+
+
+def append_patch(patch):
+    global patches
+    if not patches:
+        patches = list()
+    patches.append(patch)
+
+
+def serialize(obj):
+    serialized_obj = io.BytesIO()
+    pickle.dump(obj, serialized_obj)
+    serialized_obj.seek(0)
+    return serialized_obj
+
+
+class SpackTestProcess(object):
+    def __init__(self, fn):
+        self.fn = fn
+
+    def _restore_and_run(self, fn, test_state):
+        test_state.restore()
+        fn()
+
+    def create(self):
+        test_state = TransmitTestState()
+        return multiprocessing.Process(
+            target=self._restore_and_run,
+            args=(self.fn, test_state))
+
+
+class TransmitTestState(object):
+    _serialize = sys.version_info >= (3, 8) and sys.platform == 'darwin'
+
+    def __init__(self):
+        if TransmitTestState._serialize:
```

Comment:
Can't this use `self._serialize`?

Reply:
It could but I prefer to access class attributes via the class: I think it's useful to remind readers that an attribute is a class attribute whenever possible (since the semantics of interacting with it are different vs. an instance attribute.

Reply:
Oh interesting. I disagree, because I think class inheritance is something readers understand easily, and using the instance to access the class makes obvious that it is from its own class -- when I read this code, the first thing I thought was that this was a method from a class that the object did not belong to.

Reply:
For now this has been sidestepped by promoting `_serialize` to a module level variable (not directly because of this conversation but rather because we spoke elsewhere about aggregating the package/other-state serialization).

> using the instance to access the class makes obvious that it is from its own class

That's true, but I think a user has more to lose when they mistake a class variable for an instance variable (more likely to set it) vs. when they think a class variable belongs to another class (generally I assume folks will be more careful with setting class variables, regardless of whether they belong to the class which self is an instance of or not.